File: packages/guards/src/case-validation.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

type CaseType = 'uppercase' | 'lowercase' | 'title-case' | 'sentence-case';

interface CaseValidationOptions {
  action: 'block' | 'warn';
  expectedCase: CaseType;
}

function checkCase(text: string, expected: CaseType): boolean {
  const stripped = text.replace(/[^a-zA-Z\s]/g, '').trim();
  if (stripped.length === 0) return true;

  switch (expected) {
    case 'uppercase':
      return stripped === stripped.toUpperCase();
    case 'lowercase':
      return stripped === stripped.toLowerCase();
    case 'title-case': {
      const words = stripped.split(/\s+/);
      return words.every((w) => w.length === 0 || w[0] === w[0].toUpperCase());
    }
    case 'sentence-case': {
      const sentences = text.split(/[.!?]+/).filter((s) => s.trim().length > 0);
      return sentences.every((s) => {
        const trimmed = s.trim();
        if (trimmed.length === 0) return true;
        return trimmed[0] === trimmed[0].toUpperCase();
      });
    }
  }
}

export function caseValidation(options: CaseValidationOptions): Guard {
  return {
    name: 'case-validation',
    version: '0.1.0',
    description: 'Validate text case format (upper/lower/title/sentence)',
    category: 'format',
    supportedStages: ['output'],
    async check(
      text: string,
      _ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      const isValid = checkCase(text, options.expectedCase);

      return {
        guardName: 'case-validation',
        passed: isValid,
        action: isValid ? 'allow' : options.action,
        latencyMs: Math.round(performance.now() - start),
        details: isValid
          ? undefined
          : { expected: options.expectedCase },
      };
    },
  };
}

File: python/open_guardrail/guards/case_validation.py
"""Validate text follows expected case."""

import re
import time

from open_guardrail.core import GuardResult


def _is_upper(text: str) -> bool:
    letters = re.findall(r"[a-zA-Z]", text)
    if not letters:
        return True
    return all(c.isupper() for c in letters)


def _is_lower(text: str) -> bool:
    letters = re.findall(r"[a-zA-Z]", text)
    if not letters:
        return True
    return all(c.islower() for c in letters)


def _is_title(text: str) -> bool:
    words = re.findall(r"[a-zA-Z]+", text)
    if not words:
        return True
    return all(w[0].isupper() for w in words)


def _is_sentence(text: str) -> bool:
    sentences = re.split(r"[.!?]\s+", text.strip())
    for s in sentences:
        s = s.strip()
        if not s:
            continue
        first = re.search(r"[a-zA-Z]", s)
        if first and not first.group().isupper():
            return False
    return True


class _CaseValidation:
    def __init__(
        self,
        *,
        action: str = "warn",
        expected: str = "lower",
    ) -> None:
        self.name = "case-validation"
        self.action = action
        self.expected = expected

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        validators = {
            "upper": _is_upper,
            "lower": _is_lower,
            "title": _is_title,
            "sentence": _is_sentence,
        }
        validator = validators.get(
            self.expected, lambda t: True
        )
        valid = validator(text)
        triggered = not valid
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="case-validation",
            passed=not triggered,
            action=(
                self.action if triggered else "allow"
            ),
            message=(
                f"Text does not match {self.expected} case"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"expected": self.expected}
                if triggered
                else None
            ),
        )


def case_validation(
    *,
    action: str = "warn",
    expected: str = "lower",
) -> _CaseValidation:
    return _CaseValidation(
        action=action, expected=expected
    )
